fsw/src/crypto.c

```diff
@@ -2647,46 +2648,252 @@ static int32 Crypto_PDU(char* ingest,TC_t* tc_frame)
 {
     // Local Variables
     int32 status = OS_SUCCESS;
-    unsigned char* tc_ingest = *ingest;
 
     #ifdef DEBUG
         OS_printf(KYEL "\n----- Crypto_TC_ApplySecurity START -----\n" RESET);
     #endif
 
-    // TODO: This whole function!
-    //len_ingest = len_ingest;
-    //ingest[0] = ingest[0];
-
-    int security_header_bytes = 18;
-    int security_trailer_bytes = 16;
-    int tc_size = *len_ingest + security_header_bytes + security_trailer_bytes;
+    #ifdef DEBUG
+        OS_printf("%d TF Bytes received\n", in_frame_length);
+        OS_printf("DEBUG - ");
+        for(int i=0; i < in_frame_length; i++)
+        {
+            OS_printf("%02X", ((uint8 *)&*in_frame)[i]);
+        }
+        OS_printf("\nPrinted %d bytes\n", in_frame_length);
+    #endif
 
-    unsigned char * tempTC=NULL;
-    tempTC = (unsigned char *)malloc(tc_size * sizeof (unsigned char));
-    CFE_PSP_MemSet(tempTC, 0, tc_size);
 
-    int count = 0;
-    //Create Security Header
-    for (;count < security_header_bytes;count++){
-        tempTC[count]= 0x55; //put dummy filler bits in security header for now.
+    TC_FramePrimaryHeader_t temp_tc_header;
+    // Primary Header
+    temp_tc_header.tfvn   = ((uint8)in_frame[0] & 0xC0) >> 6;
+    temp_tc_header.bypass = ((uint8)in_frame[0] & 0x20) >> 5;
+    temp_tc_header.cc     = ((uint8)in_frame[0] & 0x10) >> 4;
+    temp_tc_header.spare  = ((uint8)in_frame[0] & 0x0C) >> 2;
+    temp_tc_header.scid   = ((uint8)in_frame[0] & 0x03) << 8;
+    temp_tc_header.scid   = temp_tc_header.scid | (uint8)in_frame[1];
+    temp_tc_header.vcid   = ((uint8)in_frame[2] & 0xFC) >> 2;
+    temp_tc_header.fl     = ((uint8)in_frame[2] & 0x03) << 8;
+    temp_tc_header.fl     = temp_tc_header.fl | (uint8)in_frame[3];
+    temp_tc_header.fsn	  = (uint8)in_frame[4];
+
+    // Check if command frame flag set
+    if (temp_tc_header.cc == 1)
+    {
+        /*
+        ** CCSDS 232.0-B-3
+        ** Section 6.3.1
+        ** "Type-C frames do not have the Security Header and Security Trailer."
+        */
+        #ifdef TC_DEBUG
+            OS_printf(KYEL "DEBUG - Received Control/Command frame - nothing to do.\n" RESET);
+        #endif
     }
 
-    //Create Frame Body
-    CFE_PSP_MemCpy(&tempTC[security_header_bytes],&tc_ingest[0],*len_ingest);
-    count+=*len_ingest;
+    // TODO: If command frame flag not set, need to know SDLS parameters
+    // What is the best way to do this - copy in the entire SA for a channel?
+    // Expect these calls will return an error is the SA is not usable
+    // Will need to know lengths of various parameters from the SA DB
+    else
+    {
+        // Query SA DB for active SA / SDLS paremeters
+        // TODO: Likely API call
+        // TODO: Magic to make sure we're getting the correct SA.. 
+        // currently SA DB allows for more than one
+        SecurityAssociation_t temp_SA;
+        int16_t spi = -1;
+        for (int i=0; i < NUM_SA;i++)
```

Comment:
Some slight adjustments needed here -- we can't just search through every SA since each SA doesn't apply to the current frame... It should do the following: 

1. First we need to determine all the SA's that match the given GVCID (TFVN, SCID, VCID) 
2. Cycle through those SAs with that GVCID until an operational one is found (error if not found)
3. Use it.

Reply:
I generally agree with this. But not sure if the logic happens here or in future SA API?